File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/action/DropGameAction.java
package org.emulinker.kaillera.controller.v086.action;

import com.google.common.flogger.FluentLogger;
import org.emulinker.kaillera.controller.messaging.MessageFormatException;
import org.emulinker.kaillera.controller.v086.V086Controller;
import org.emulinker.kaillera.controller.v086.protocol.*;
import org.emulinker.kaillera.model.*;
import org.emulinker.kaillera.model.event.*;
import org.emulinker.kaillera.model.exception.DropGameException;

public class DropGameAction implements V086Action, V086GameEventHandler {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private static final String desc = "DropGameAction";
  private static DropGameAction singleton = new DropGameAction();

  public static DropGameAction getInstance() {
    return singleton;
  }

  private int actionCount = 0;
  private int handledCount = 0;

  private DropGameAction() {}

  @Override
  public int getActionPerformedCount() {
    return actionCount;
  }

  @Override
  public int getHandledEventCount() {
    return handledCount;
  }

  @Override
  public String toString() {
    return desc;
  }

  @Override
  public void performAction(V086Message message, V086Controller.V086ClientHandler clientHandler)
      throws FatalActionException {
    if (!(message instanceof PlayerDrop_Request))
      throw new FatalActionException("Received incorrect instance of PlayerDrop: " + message);

    actionCount++;

    try {
      clientHandler.getUser().dropGame();
    } catch (DropGameException e) {
      logger.atFine().withCause(e).log("Failed to drop game");
    }
  }

  @Override
  public void handleEvent(GameEvent event, V086Controller.V086ClientHandler clientHandler) {
    handledCount++;

    UserDroppedGameEvent userDroppedEvent = (UserDroppedGameEvent) event;

    try {
      KailleraUser user = userDroppedEvent.getUser();
      int playerNumber = userDroppedEvent.getPlayerNumber();
      //			clientHandler.send(PlayerDrop_Notification.create(clientHandler.getNextMessageNumber(),
      // user.getName(), (byte) game.getPlayerNumber(user)));
      if (user.getStealth() == false)
        clientHandler.send(
            PlayerDrop_Notification.create(
                clientHandler.getNextMessageNumber(), user.getName(), (byte) playerNumber));
    } catch (MessageFormatException e) {
      logger.atSevere().withCause(e).log("Failed to contruct PlayerDrop_Notification message");
    }
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/action/DropGameAction.kt
package org.emulinker.kaillera.controller.v086.action

import com.google.common.flogger.FluentLogger
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.v086.V086ClientHandler
import org.emulinker.kaillera.controller.v086.protocol.PlayerDropNotification
import org.emulinker.kaillera.controller.v086.protocol.PlayerDropRequest
import org.emulinker.kaillera.model.event.UserDroppedGameEvent
import org.emulinker.kaillera.model.exception.DropGameException

class DropGameAction : V086Action<PlayerDropRequest>, V086GameEventHandler<UserDroppedGameEvent> {
  override fun toString() = "DropGameAction"

  @Throws(FatalActionException::class)
  override fun performAction(message: PlayerDropRequest, clientHandler: V086ClientHandler) {
    try {
      clientHandler.user.dropGame()
    } catch (e: DropGameException) {
      logger.atFine().withCause(e).log("Failed to drop game")
    }
  }

  override fun handleEvent(event: UserDroppedGameEvent, clientHandler: V086ClientHandler) {
    try {
      val user = event.user
      val playerNumber = event.playerNumber
      //			clientHandler.send(PlayerDrop.Notification.create(clientHandler.getNextMessageNumber(),
      // user.getName(), (byte) game.getPlayerNumber(user)));
      if (!user.inStealthMode)
        clientHandler.send(PlayerDropNotification(0, user.name!!, playerNumber.toByte()))
    } catch (e: MessageFormatException) {
      logger.atSevere().withCause(e).log("Failed to construct PlayerDrop.Notification message")
    }
  }

  companion object {
    private val logger = FluentLogger.forEnclosingClass()
  }
}
